llama_index/llm_predictor/base.py

```diff
@@ -132,29 +131,128 @@ def _log_template_data(
+                prompt=PromptTemplate(template=prompt, prompt_type=PromptType.SUMMARY),
+            )
+            chat_response = program()
+            return chat_response.json()
+
+        if self._llm.metadata.is_chat_model and messages:
+            try:
+                from llama_index.program.openai_program import OpenAIPydanticProgram
+
+                program = OpenAIPydanticProgram.from_defaults(
+                    output_cls=output_cls,
+                    llm=self._llm,
+                    messages=messages,
+                )
+            except ValueError:
+                from llama_index.program.llm_program import LLMTextCompletionProgram
+
+                program = LLMTextCompletionProgram.from_defaults(
+                    output_parser=PydanticOutputParser(output_cls=output_cls),
+                    messages=messages,
+                    llm=self._llm,
+                )
+            chat_response = program()
+            return chat_response.json()
+        return "Empty Response"
+
+    async def _arun_program(
+        self,
+        output_cls: BaseModel,
+        prompt: Optional[str] = None,
+        messages: Optional[List[ChatMessage]] = None,
+    ) -> str:
+        program: BasePydanticProgram
+        if not self._llm.metadata.is_chat_model and prompt:
+            from llama_index.program.llm_program import LLMTextCompletionProgram
+
+            program = LLMTextCompletionProgram.from_defaults(
+                output_parser=PydanticOutputParser(output_cls=output_cls),
+                llm=self._llm,
+                prompt=PromptTemplate(template=prompt, prompt_type=PromptType.SUMMARY),
+            )
+            chat_response = await program.acall()
+            return chat_response.json()
+
+        if self._llm.metadata.is_chat_model and messages:
+            try:
+                from llama_index.program.openai_program import OpenAIPydanticProgram
+
+                program = OpenAIPydanticProgram.from_defaults(
+                    output_cls=output_cls,
+                    llm=self._llm,
+                    messages=messages,
+                )
+            except ValueError:
+                from llama_index.program.llm_program import LLMTextCompletionProgram
+
+                program = LLMTextCompletionProgram.from_defaults(
+                    output_parser=PydanticOutputParser(output_cls=output_cls),
+                    messages=messages,
+                    llm=self._llm,
+                )
+            chat_response = await program.acall()
+            return chat_response.json()
+        return "Empty Response"
+
+    def predict(
+        self,
+        prompt: BasePromptTemplate,
+        output_cls: Optional[BaseModel] = None,
+        **prompt_args: Any
+    ) -> str:
         """Predict."""
         self._log_template_data(prompt, **prompt_args)
 
         if self._llm.metadata.is_chat_model:
             messages = prompt.format_messages(llm=self._llm, **prompt_args)
```

Comment:
at a high-level, instead of formatting the prompt template here and passing it thorugh to run_program, is it possible to pass the prompt template + kwargs down to run_program? seems more consistent with how pydantic program is used

Reply:
Fixed!